From 891b8e27b7558fd82fe07dfe51e21cf06178cc28 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tor=20Arne=20Vestb=C3=B8?= <tor.arne.vestbo@qt.io>
Date: Mon, 29 Sep 2025 14:00:22 +0200
Subject: [PATCH 251/553] Add QCoreTextFontEngine::glyphCount override

The base QFontEngine::glyphCount was reporting 0 glyphs. Possibly
due to not working on instances of QFontEngineMulti. But in any case
it makes sense to use the CoreText implementation if we can.

Pick-to: 6.8 6.5
Change-Id: I267b699b7cb71a7fc6758a375b783044a0b618b0
Reviewed-by: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
(cherry picked from commit 6fc0f4ccb4a90a29c2af89ffd127ec38fd4e7ac9)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/text/coretext/qfontengine_coretext.mm  | 5 +++++
 src/gui/text/coretext/qfontengine_coretext_p.h | 1 +
 src/gui/text/qfontengine.cpp                   | 5 +++++
 src/gui/text/qfontengine_p.h                   | 1 +
 4 files changed, 12 insertions(+)

diff --git a/src/gui/text/coretext/qfontengine_coretext.mm b/src/gui/text/coretext/qfontengine_coretext.mm
index 14173e0ac6c..619d4fabcf1 100644
--- a/src/gui/text/coretext/qfontengine_coretext.mm
+++ b/src/gui/text/coretext/qfontengine_coretext.mm
@@ -301,6 +301,11 @@ void QCoreTextFontEngine::init()
     }
 }
 
+int QCoreTextFontEngine::glyphCount() const
+{
+    return CTFontGetGlyphCount(ctfont);
+}
+
 glyph_t QCoreTextFontEngine::glyphIndex(uint ucs4) const
 {
     int len = 0;
diff --git a/src/gui/text/coretext/qfontengine_coretext_p.h b/src/gui/text/coretext/qfontengine_coretext_p.h
index fe4d2c6cc56..24cb29c7ce8 100644
--- a/src/gui/text/coretext/qfontengine_coretext_p.h
+++ b/src/gui/text/coretext/qfontengine_coretext_p.h
@@ -37,6 +37,7 @@ public:
     QCoreTextFontEngine(CGFontRef font, const QFontDef &def);
     ~QCoreTextFontEngine();
 
+    int glyphCount() const override;
     glyph_t glyphIndex(uint ucs4) const override;
     QString glyphName(glyph_t index) const override;
     glyph_t findGlyph(QLatin1StringView name) const override;
diff --git a/src/gui/text/qfontengine.cpp b/src/gui/text/qfontengine.cpp
index dd58975bc93..0ad45b1f280 100644
--- a/src/gui/text/qfontengine.cpp
+++ b/src/gui/text/qfontengine.cpp
@@ -1933,6 +1933,11 @@ QFontEngine *QFontEngineMulti::loadEngine(int at)
     return nullptr;
 }
 
+int QFontEngineMulti::glyphCount() const
+{
+    return engine(0)->glyphCount();
+}
+
 glyph_t QFontEngineMulti::glyphIndex(uint ucs4) const
 {
     glyph_t glyph = engine(0)->glyphIndex(ucs4);
diff --git a/src/gui/text/qfontengine_p.h b/src/gui/text/qfontengine_p.h
index 5e2c2963ca4..160652e9c74 100644
--- a/src/gui/text/qfontengine_p.h
+++ b/src/gui/text/qfontengine_p.h
@@ -453,6 +453,7 @@ public:
     explicit QFontEngineMulti(QFontEngine *engine, int script, const QStringList &fallbackFamilies = QStringList());
     ~QFontEngineMulti();
 
+    virtual int glyphCount() const override;
     virtual glyph_t glyphIndex(uint ucs4) const override;
     virtual int stringToCMap(const QChar *str, int len, QGlyphLayout *glyphs, int *nglyphs, ShaperFlags flags) const override;
 
-- 
2.51.2

